WIP: presence event engine

mohitpubnub opened the pull request:
implementation for presence event engine

parfeon commented on src/event-engine/states/handshake_failure.ts:
```diff
@@ -1,9 +1,10 @@
 import { State } from '../core/state';
```
`HANDSHAKE_FAILED` is missing handling for `SUBSCRIPTION_CHANGED`, `SUBSCRIPTION_RESTORED` and `UNSUBSCRIBE_ALL`.

mohitpubnub replied:
Added all 3 ✔️ 

https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/event-engine/states/handshake_failed.ts#L18

parfeon commented on src/event-engine/presence/states/heartbeating.ts:
```diff
@@ -0,0 +1,68 @@
+      channels: context.channels,
+      groups: context.groups,
+    },
+    [emitStatus(event.payload)],
```
Do we have any status here right now? Or is it for backward compatibility?

mohitpubnub replied:
removed as per new specs ✔️

parfeon commented on src/event-engine/presence/states/heartbeat_stopped.ts:
```diff
@@ -0,0 +1,50 @@
+);
+
+HeartbeatStoppedState.on(leftAll.type, (context, _) =>
+  HeartbeatInactiveState.with(undefined, [leave(context.channels, context.groups)]),
```
Same situation with `leave` effect as in response to `LEFT` event handling.

mohitpubnub replied:
Agree. 
removed ✔️

parfeon commented on src/event-engine/presence/states/heartbeat_cooldown.ts:
```diff
@@ -0,0 +1,54 @@
+  groups: string[];
+};
+
+export const HeartbeatCooldownState = new State<HeartbeatCooldownStateContext, Events, Effects>('HEARTBEATCOOLDOWN');
```
Will it be better to use `HEARTBEAT_COOLDOWN`?

mohitpubnub replied:
Renamed ✔️ 
https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/event-engine/presence/states/heartbeat_cooldown.ts#L13

parfeon commented on src/event-engine/presence/dispatcher.ts:
```diff
@@ -0,0 +1,108 @@
+          const result = await heartbeat({
+            channels: payload.channels,
+            channelGroups: payload.groups,
+            state: presenceState,
```
Should we send it only if maintain state flag is set to `true`?

parfeon commented on src/event-engine/presence/dispatcher.ts:
```diff
@@ -0,0 +1,108 @@
+            const result = await heartbeat({
+              channels: payload.channels,
+              channelGroups: payload.groups,
+              state: presenceState,
```
Should this to be set if the corresponding configuration flag is set to `true`?

parfeon commented on src/event-engine/dispatcher.ts:
```diff
@@ -70,77 +77,87 @@ export class EventEngineDispatcher extends Dispatcher<effects.Effects, Dependenc
+              channels: payload.channels,
+              channelGroups: payload.groups,
+              filterExpression: config.filterExpression,
+              state: presenceState,
```
Can we use the `config.maintainPresenceState` flag and don't pass `presenceState` into `handshake`?

mohitpubnub replied:
it's conditional now ✔️ 
same as with heartbeat

https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/event-engine/dispatcher.ts#L37

parfeon reviewed (changes requested):
Mostly it is in good shape!

parfeon commented on src/event-engine/core/retryPolicy.ts:
```diff
@@ -0,0 +1,52 @@
+      },
+
+      getDelay(attempt: number) {
+        const calculatedDelay = Math.trunc(Math.pow(2, attempt)) * 1000 + Math.random() * 1000;
```
@mohitpubnub what value will be for `attempt` when it called first time? Per specs, it should be 0.

mohitpubnub replied:
It starts from 0. Whenever it enters *Reconnecting state.

parfeon reviewed (commented):
One more

mohitpubnub commented:
Stats:

- [x] Naming conventions [event engine states, event names, config helper methods]
- [x] Missing transitions [specifically for `*FAILED` states, and `reconnect` or `restore`]
- [x] Removed EmitStatus for presence [previously added for backward compatibility]
- [x] SetState [`state` param passed conditionally when `maintainPresenceState` set]
- [x] Naming conventions for presence event engine. [heartbeat cooldown]
- [x] RetryConfiguration [added randomness for linear policy, array for non retrying http status code]
- [x] removed unnecessary left event handling wherever not applicable

mohitpubnub commented on src/core/components/config.js:
```diff
@@ -183,8 +187,8 @@ export default class {
 
-    // flag for beta subscribe feature enablement
-    this.enableSubscribeBeta = setup.enableSubscribeBeta ?? false;
+    this.enableEventEngine = setup.enableEventEngine ?? false;
+    this.maintainPresenceState = setup.maintainPresenceState ?? true;
```
As per discussion keeping it like default not set for ee.
If ee set then presence maintained by default as per behaviour specification, knowing that enabling ee may trigger some changes.

mohitpubnub commented on src/core/components/config.js:
```diff
@@ -353,6 +361,21 @@ export default class {
     return '7.4.5';
   }
 
+  setRetryConfiguration(configuration) {
```
✔️ 
https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/core/components/config.js#L364

mohitpubnub commented on src/core/endpoints/subscriptionUtils/handshake.js:
```diff
@@ -25,6 +25,13 @@ const endpoint = {
       outParams['channel-group'] = params.channelGroups.join(',');
     }
     outParams.tt = 0;
+    if (params.state) {
```
updated in effect implementation.

this will stay as when `state` passed then we need to check and stringify it for http request query param

mohitpubnub commented on src/event-engine/core/retryPolicy.ts:
```diff
@@ -0,0 +1,52 @@
+      delay: configuration.delay,
+      maximumRetry: configuration.maximumRetry,
+      shouldRetry(error: any, attempt: number) {
+        if (error?.status?.statusCode === 403) {
```
updated:
https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/event-engine/core/retryPolicy.ts#L8

We can extend array as per requirement.

Retry-After will be handled here once we start getting it

mohitpubnub commented on src/event-engine/core/retryPolicy.ts:
```diff
@@ -0,0 +1,52 @@
+      getDelay(attempt: number) {
+        const calculatedDelay = Math.trunc(Math.pow(2, attempt)) * 1000 + Math.random() * 1000;
+        if (calculatedDelay > 150000) {
+          return 150000;
```
✔️ 
https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/event-engine/core/retryPolicy.ts#L44
Defaulting to user defined maximum

mohitpubnub commented on src/event-engine/core/retryPolicy.ts:
```diff
@@ -0,0 +1,52 @@
+        return this.maximumRetry > attempt;
+      },
+      getDelay(_: number) {
+        return this.delay * 1000;
```
✔️ 
https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/event-engine/core/retryPolicy.ts#L14
Now it put additional random delay

mohitpubnub commented on src/event-engine/dispatcher.ts:
```diff
@@ -18,17 +24,18 @@ export class EventEngineDispatcher extends Dispatcher<effects.Effects, Dependenc
     this.on(
       effects.handshake.type,
-      asyncHandler(async (payload, abortSignal, { handshake }) => {
+      asyncHandler(async (payload, abortSignal, { handshake, presenceState, config }) => {
```
✔️ 
https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/event-engine/dispatcher.ts#L37

effect will pass `state` based on `maintainPresenceState` configuration value

mohitpubnub commented on src/event-engine/dispatcher.ts:
```diff
@@ -70,77 +77,87 @@ export class EventEngineDispatcher extends Dispatcher<effects.Effects, Dependenc
+        if (config.retryConfiguration && config.retryConfiguration.shouldRetry(payload.reason, payload.attempts)) {
+          abortSignal.throwIfAborted();
+
+          await delay(config.retryConfiguration.getDelay(payload.attempts));
```
✅  Agree

mohitpubnub commented on src/event-engine/dispatcher.ts:
```diff
@@ -70,77 +77,87 @@ export class EventEngineDispatcher extends Dispatcher<effects.Effects, Dependenc
-          if (error instanceof PubNubError) {
-            return engine.transition(events.reconnectingFailure(error));
+    this.on(
+      effects.reconnect.type,
```
✔️  Yes, It should be `receiveReconnect`
updated:
https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/event-engine/dispatcher.ts#L96

mohitpubnub commented on src/event-engine/effects.ts:
```diff
@@ -9,19 +9,29 @@ export const handshake = createManagedEffect('HANDSHAKE', (channels: string[], g
+  (context: ReceiveReconnectingStateContext) => context,
+);
 
 export const handshakeReconnect = createManagedEffect(
```
Agree!!
Updating it now causes missing `CANCEL_HANDSHAKE` in contract.

mohitpubnub commented on src/event-engine/states/unsubscribed.ts:
```diff
@@ -1,10 +1,22 @@
+  HandshakingState.with({
+    channels: event.payload.channels,
+    groups: event.payload.groups,
+    timetoken: event.payload.timetoken,
```
✔️ 
It's clean now,
https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/event-engine/states/unsubscribed.ts#L8

mohitpubnub commented on src/event-engine/states/receiving.ts:
```diff
@@ -1,10 +1,11 @@
 import { State } from '../core/state';
```
✔️ 
Added all missing transition.  Completely missed by following older specification.

mohitpubnub commented on src/event-engine/states/receive_stopped.ts:
```diff
@@ -13,11 +15,26 @@ type ReceiveStoppedStateContext = {
+  ReceivingState.with({
+    channels: event.payload.channels,
+    groups: event.payload.groups,
+    cursor: context.cursor,
```
Handled in updated transitions definitions. ✔️

mohitpubnub commented on src/event-engine/states/receive_stopped.ts:
```diff
@@ -13,11 +15,26 @@ type ReceiveStoppedStateContext = {
+    channels: context.channels,
+    groups: context.groups,
+  }),
+);
```
Updated ✔️

mohitpubnub commented on src/event-engine/states/receive_failure.ts:
```diff
@@ -14,12 +16,13 @@ export type ReceiveFailureStateContext = {
 };
 
-export const ReceiveFailureState = new State<ReceiveFailureStateContext, Events, Effects>('RECEIVE_FAILURE');
+export const ReceiveFailureState = new State<ReceiveFailureStateContext, Events, Effects>('RECEIVE_FAILED');
```
Renamed ✔️ 
https://github.com/pubnub/javascript/blob/presence/src/event-engine/states/receive_failed.ts

mohitpubnub commented on src/event-engine/states/receive_failure.ts:
```diff
@@ -14,12 +16,13 @@ export type ReceiveFailureStateContext = {
+  HandshakingState.with({
+    channels: context.channels,
+    groups: context.groups,
+    timetoken: context.cursor.timetoken,
```
Handled ✔️ 
https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/event-engine/states/receive_failed.ts#L24

mohitpubnub commented on src/event-engine/states/receive_failure.ts:
```diff
@@ -30,3 +33,21 @@ ReceiveFailureState.on(disconnect.type, (context) =>
+  HandshakingState.with({
+    channels: event.payload.channels,
+    groups: event.payload.groups,
+    timetoken: event.payload.timetoken,
```
✔️

mohitpubnub commented on src/event-engine/states/handshake_stopped.ts:
```diff
@@ -8,10 +8,10 @@ type HandshakeStoppedStateContext = {
 };
 
-export const HandshakeStoppedState = new State<HandshakeStoppedStateContext, Events, Effects>('STOPPED');
+export const HandshakeStoppedState = new State<HandshakeStoppedStateContext, Events, Effects>('HANDSHAKE_STOPPED');
```
Added 
https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/event-engine/states/handshake_stopped.ts#L34
✔️

mohitpubnub commented on src/event-engine/states/handshake_reconnecting.ts:
```diff
@@ -19,34 +32,53 @@ export const HandshakeReconnectingState = new State<HandshakeReconnectingStateCo
-HandshakeReconnectingState.on(reconnectingSuccess.type, (context, event) =>
-  ReceivingState.with(
+HandshakeReconnectingState.on(handshakingReconnectingSuccess.type, (context, event) => {
+  const cursor = context.timetoken ? { timetoken: context.timetoken, region: 1 } : event.payload.cursor;
```
✔️ 
Using received region..
https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/event-engine/states/handshake_reconnecting.ts#L38

mohitpubnub commented on src/event-engine/states/handshake_reconnecting.ts:
```diff
@@ -19,34 +32,53 @@ export const HandshakeReconnectingState = new State<HandshakeReconnectingStateCo
+    {
+      channels: context.channels,
+      groups: context.groups,
+    },
```
✔️ 
User provided information carry forwarded
https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/event-engine/states/handshake_reconnecting.ts#L73

mohitpubnub commented on src/event-engine/states/handshake_reconnecting.ts:
```diff
@@ -19,34 +32,53 @@ export const HandshakeReconnectingState = new State<HandshakeReconnectingStateCo
+      groups: context.groups,
+      channels: context.channels,
+      reason: context.reason,
+    },
```
Agree, Now `handshake_reconnecting` state carry forward context information to transitioned state

https://github.com/pubnub/javascript/blob/presence/src/event-engine/states/handshake_reconnecting.ts#L73

mohitpubnub commented on src/event-engine/states/handshake_reconnecting.ts:
```diff
@@ -19,34 +32,53 @@ export const HandshakeReconnectingState = new State<HandshakeReconnectingStateCo
+);
+
+HandshakeReconnectingState.on(subscriptionChange.type, (_, event) =>
+  HandshakingState.with({ channels: event.payload.channels, groups: event.payload.groups }),
```
✔️  updated

mohitpubnub commented on src/event-engine/states/handshake_reconnecting.ts:
```diff
@@ -19,34 +32,53 @@ export const HandshakeReconnectingState = new State<HandshakeReconnectingStateCo
+);
+
+HandshakeReconnectingState.on(restore.type, (_, event) =>
+  HandshakingState.with({ channels: event.payload.channels, groups: event.payload.groups }),
```
✔️

mohitpubnub commented on src/event-engine/states/handshake_failure.ts:
```diff
@@ -14,16 +15,11 @@ export type HandshakeFailureStateContext = {
 
 export const HandshakeFailureState = new State<HandshakeFailureStateContext, Events, Effects>('HANDSHAKE_FAILURE');
```
aligned ✔️

mohitpubnub commented on src/event-engine/presence/states/heartbeat_stopped.ts:
```diff
@@ -0,0 +1,50 @@
+      channels: context.channels.filter((channel) => !event.payload.channels.includes(channel)),
+      groups: context.groups.filter((group) => !event.payload.groups.includes(group)),
+    },
+    [leave(event.payload.channels, event.payload.groups)],
```
Agree. Not needed here so removed ✔️

mohitpubnub commented on src/event-engine/presence/states/heartbeat_stopped.ts:
```diff
@@ -0,0 +1,50 @@
+    [leave(context.channels, context.groups)],
+  ),
+);
+
```
removed ✔️ 

https://github.com/pubnub/javascript/blob/presence/src/event-engine/presence/states/heartbeat_stopped.ts

mohitpubnub commented on src/event-engine/presence/states/heartbeat_inactive.ts:
```diff
@@ -0,0 +1,15 @@
+  }),
+);
+
+HeartbeatInactiveState.on(left.type, (_, event) => HeartbeatInactiveState.with());
```
removed as its not in specification ✔️

parfeon commented on src/event-engine/states/receive_failed.ts:
```diff
@@ -0,0 +1,45 @@
+  HandshakingState.with({
+    channels: event.payload.channels,
+    groups: event.payload.groups,
+    cursor: { timetoken: event.payload.timetoken, region: event.payload?.region ?? 0 },
```
In other states, we used here region from the context instead of `0`.

mohitpubnub replied:
Updating, 

The reason was:
here, It's getting into Handshaking.
So,
Once it get successfull response for handshake, It will anyway use server provided region (along with user provided tt) with the thought that it's recommended to use server provided region value

parfeon replied:
Ok. Got the point. Let it be `0` :)

mohitpubnub replied:
unified across the state: and when values are not provided by user and also state context doesn't have it.. Then only it will default to 0 which indicates falsy value while handshaking

parfeon commented on src/event-engine/states/receive_failed.ts:
```diff
@@ -0,0 +1,45 @@
+  HandshakingState.with({
+    channels: context.channels,
+    groups: context.groups,
+    cursor: {
+      timetoken: event.payload?.timetoken ?? '0',
+      region: event.payload?.region ?? 0,
+    },
```
Should we check the `context` for cursor information instead of defaulting to `0`?

mohitpubnub replied:
unified across the states

parfeon commented on src/event-engine/states/handshaking.ts:
```diff
@@ -24,25 +36,50 @@ HandshakingState.on(subscriptionChange.type, (context, event) => {
+      groups: context.groups,
+      cursor: {
+        timetoken: context.cursor?.timetoken ?? event.payload.timetoken,
+        region: event.payload.region,
```
Is it possible that `region` is known in the `context`? If it is, then it will be better to check conditionally.

mohitpubnub replied:
It's handshake success... So now we are using region provided by server and disregarding custom region. As per design description.

parfeon commented on src/event-engine/states/handshaking.ts:
```diff
@@ -24,25 +36,50 @@ HandshakingState.on(subscriptionChange.type, (context, event) => {
+  return HandshakeReconnectingState.with({
+    channels: context.channels,
+    groups: context.groups,
+    timetoken: context.cursor?.timetoken,
```
All events in our specs use the `cursor` definition, which is both `timetoken` and `region` in one variable. Can we sync with specs in all places, where `timetoken` can be replaced with `cursor`? 
In this code, we're throwing off `region` but it is important in this case.

mohitpubnub replied:
unified ✔️

parfeon commented on src/event-engine/states/handshake_stopped.ts:
```diff
@@ -1,20 +1,45 @@
+    ...context,
+    cursor: {
+      timetoken: event.payload?.timetoken ?? context.cursor?.timetoken ?? '0',
+      region: event.payload?.region ?? 0,
```
Should we check `region` in `context` in the same way as with `timetoken` to make sure that we avoid any issues because of old `region`?

mohitpubnub replied:
updated: and defaulting to 0 only when event/context both doesn't have cursor info with region

parfeon commented on src/event-engine/states/handshake_reconnecting.ts:
```diff
@@ -19,34 +32,63 @@ export const HandshakeReconnectingState = new State<HandshakeReconnectingStateCo
   HandshakeStoppedState.with({
     channels: context.channels,
     groups: context.groups,
+    cursor: {
+      timetoken: context.timetoken ?? '0',
+      region: 0,
```
Do we really need to strip away `region`?

mohitpubnub replied:
now it access incoming/context value.

parfeon commented on src/event-engine/states/handshake_failed.ts:
```diff
@@ -0,0 +1,44 @@
+    groups: context.groups,
+    cursor: {
+      timetoken: event.payload?.timetoken ?? context.timetoken ?? '0',
+      region: event.payload?.region ?? 0,
```
Is there a reason to ignore `region` from the `context`?

mohitpubnub replied:
updated.

mohitpubnub commented on src/event-engine/states/receiving.ts:
```diff
@@ -14,33 +23,61 @@ export type ReceivingStateContext = {
+    channels: event.payload.channels,
+    groups: event.payload.groups,
+    cursor: {
+      timetoken: event.payload.timetoken,
```
> Should we change the type of timetoken field in event

Didn't get the point. Do we require changing the type of `timetoken` value anywhere?

parfeon replied:
In our specs, all events operate on `cursor`, not `timetoken`. So I was curious on whether we should sync it somehow with all states and events?

mohitpubnub replied:
I get it..  On it.

mohitpubnub replied:
unified across the state:

mohitpubnub commented on src/event-engine/core/retryPolicy.ts:
```diff
@@ -1,16 +1,26 @@
       maximumRetry: configuration.maximumRetry,
       shouldRetry(error: any, attempt: number) {
-        if (error?.status?.statusCode === 403) {
+        if (RetryPolicy.excludedErrorCodes.includes(error?.status?.statusCode)) {
```
What about  accepting additional params(the value comming from failure Reason) which will override calculated one
`getDelay(attempt, retryAfter)`

parfeon replied:
Sounds good :)

mohitpubnub replied:
now, `getDelay` has reference of error details
so once we have additional header details it will propagate through error

parfeon commented on src/event-engine/states/handshake_stopped.ts:
```diff
@@ -1,20 +1,42 @@
+HandshakeStoppedState.on(reconnect.type, (context, event) =>
+  HandshakingState.with({
+    ...context,
+    cursor: context.cursor || event.payload.cursor,
```
Should we use this version, try to handle user-provided `cursor` before trying checking `cursor` from the context
```suggestion
    cursor: event.payload.cursor || context.cursor,
```

mohitpubnub replied:
updated wherever applicable ✔️

parfeon commented on src/event-engine/states/handshake_reconnecting.ts:
```diff
@@ -19,34 +33,60 @@ export const HandshakeReconnectingState = new State<HandshakeReconnectingStateCo
+  }),
+);
+
+HandshakeReconnectingState.on(subscriptionChange.type, (_, event) =>
+  HandshakingState.with({ channels: event.payload.channels, groups: event.payload.groups }),
+);
```
It looks like in this case we will throw away `cursor` which user potentially could pass in the other state, and we were keeping it for transition to the `Receiving` state.

```suggestion
HandshakeReconnectingState.on(subscriptionChange.type, (context, event) =>
  HandshakingState.with({ 
    channels: event.payload.channels, 
    groups: event.payload.groups,
    cursor: context.cursor,
  }),
);
```

mohitpubnub replied:
committed changes ✔️

parfeon commented on src/event-engine/states/handshake_failed.ts:
```diff
@@ -0,0 +1,42 @@
+  HandshakingState.with({
+    channels: context.channels,
+    groups: context.groups,
+    cursor: context.cursor || event.payload.cursor,
```
Should we use this version, try to handle user-provided `cursor` before trying checking `cursor` from the context
```suggestion
    cursor: event.payload.cursor || context.cursor,
```

parfeon commented on src/event-engine/states/handshake_failed.ts:
```diff
@@ -0,0 +1,42 @@
+
+export const HandshakeFailedState = new State<HandshakeFailedStateContext, Events, Effects>('HANDSHAKE_FAILED');
+
+HandshakeFailedState.on(subscriptionChange.type, (_, event) => {
+  return HandshakingState.with({ channels: event.payload.channels, groups: event.payload.groups });
+});
```
It looks like in this case we will throw away `cursor` which user potentially could pass in the other state, and we were keeping it for transition to the `Receiving` state.

```suggestion
HandshakeFailedState.on(subscriptionChange.type, (context, event) => {
  return HandshakingState.with({ 
    channels: event.payload.channels, 
    groups: event.payload.groups,
    cursor: context.cursor,
  });
});
```

mohitpubnub replied:
✔️

mohitpubnub commented on src/event-engine/states/receive_stopped.ts:
```diff
@@ -20,4 +21,26 @@ ReceiveStoppedState.on(subscriptionChange.type, (context, event) =>
+    groups: context.groups,
+    cursor: {
+      timetoken: !!event.payload.cursor.timetoken ? event.payload.cursor?.timetoken : context.cursor.timetoken,
+      region: event.payload.cursor.region ? event.payload.cursor.region : context.cursor.region,
```
replaced for receive* states
and for handshake* states defaulting to 0 incase of possible undefined context cursor.

mohitpubnub commented on src/event-engine/core/retryPolicy.ts:
```diff
@@ -0,0 +1,77 @@
+          return this.maximumDelay;
+        } else {
+          return calculatedDelay;
+        }
```
committed suggested with `const` value.

parfeon commented on src/event-engine/core/retryPolicy.ts:
```diff
@@ -0,0 +1,77 @@
+        if (reason?.retryAfter) {
+          return reason.retryAfter * 1000;
+        }
+        return (this.delay + Math.random()) * 1000;
```
Yeah, my bad. Too much of `Rust` :D

mohitpubnub commented on src/event-engine/states/receive_failed.ts:
```diff
@@ -0,0 +1,48 @@
+  }),
+);
+
+ReceiveFailedState.on(subscriptionChange.type, (_, event) =>
+  HandshakingState.with({
+    channels: event.payload.channels,
+    groups: event.payload.groups,
+  }),
+);
```
Hmm!! 
The subscription change with only channels/groups should propagate channels/groups and override other context.
That's what my understanding.

but yeah, I think this need to carry forward, even with current behaviour It persists earlier tt.

Agree
👍🏻

mohitpubnub commented on src/event-engine/states/handshaking.ts:
```diff
@@ -24,25 +35,51 @@ HandshakingState.on(subscriptionChange.type, (context, event) => {
   return HandshakingState.with({ channels: event.payload.channels, groups: event.payload.groups });
```
I think it should carry forward possible cursor value here as well. This is unlikely to happen as You earlier mentioned somewhere that handshake is one shot. but we need to handle this here. instead losing value.

mohitpubnub replied:
@parfeon   Added optional cursor value.

parfeon reviewed (approved):
LGTM!

mohitpubnub commented:
@pubnub-release-bot release as v7.5.0

mohitpubnub commented:
@pubnub-release-bot release as v7.5.0
